File: src/main/java/it/ha/ax/laboration2website/service/VideoService.java
package it.ha.ax.laboration2website.service;

import it.ha.ax.laboration2website.entity.Video;
import it.ha.ax.laboration2website.repository.VideoRepository;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
public class VideoService {


    private final VideoRepository videoRepository;

    public VideoService(VideoRepository videoRepository){

        this.videoRepository = videoRepository;
    }

    // get all videos
    public List<Video> getAllVideos(){
        return videoRepository.findAll();
    }

    // Add a new video
    public Video createVideo(String title, String link, String addedBy, String category, String description){
        Video video = new Video(null, title, link, LocalDateTime.now(), addedBy, category, description);
        return videoRepository.save(video);
    }
    // Update an existing video
    public Video updateVideo(Long id, String title, String link, String addedBy, String category, String description){
        Optional<Video> existingVideo = videoRepository.findById(id);

        if(existingVideo.isPresent()){
            Video video = existingVideo.get();
            video.setTitle(title);
            video.setLink(link);
            video.setAddedBy(addedBy);
            video.setCategory(category);
            video.setDescription(description);
            return videoRepository.save(video);

        } else {
            throw new RuntimeException("Video not found with id: " + id);
        }

    }
    // Delete a video
    public void deleteVideo(Long id){
        if(videoRepository.existsById(id)){
            videoRepository.deleteById(id);
        } else {
            throw new RuntimeException("Video not found by id:" + id);
        }
    }
    // Retrieve videos by a specific user
    public List<Video> getVideosByUser(String username) {
        return videoRepository.findByAddedBy(username);
    }
    // Search videos by title keyword
    public List<Video> searchVideosByTitle(String keyword) {

        return videoRepository.searchByTitle(keyword);
    }
}

File: frontend-angular/src/app/video.service.ts
import { Injectable} from "@angular/core";
import {HttpClient } from "@angular/common/http";
import { Observable} from "rxjs";


export interface Video {

  id: number;
  title: string;
  link: string;
  addedBy: string;
  category: string;
  description: string;

}

@Injectable({
  providedIn: 'root'
})

export class VideoService{

  private baseUrl = 'http://localhost:8080/video';

  constructor(private http:HttpClient) {}

  // get all videos from the server
  getAllVideos(): Observable<Video[]>{
    return this.http.get<Video[]>(`${this.baseUrl}/`);
  }

  // Search for videos based on a query string
  searchVideos(query: string): Observable<Video[]> {
    return this.http.get<Video[]>(`${this.baseUrl}/search?q=${query}`);
  }

  // Retrieve all videos added by a specific user
  getVideosByUser(username: string | null): Observable<Video[]> {
    return this.http.get<Video[]>(`${this.baseUrl}/user/${username}`);
  }

  // Add a new video by sending video details to the server
  addVideo(video: {
    addedBy: string;
    link: string;
    description: string;
    title: string;
    category: string
  }): Observable<Video> {
    return this.http.post<Video>(`${this.baseUrl}/upload`, video);
  }

  // Update an existing video's details
  updateVideo(video: {
    addedBy: string;
    link: string;
    description: string;
    id: null;
    title: string;
    category: string
  }): Observable<Video> {
    return this.http.put<Video>(`${this.baseUrl}/${video.id}`, video);
  }

  // Delete a video by its ID
  deleteVideo(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

}
